third-party/rsutils/src/special-folder.cpp

```diff
@@ -0,0 +1,139 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2023 Intel Corporation. All Rights Reserved.
+
+#include <rsutils/os/special-folder.h>
+#include <rsutils/easylogging/easyloggingpp.h>
+
+#ifdef _WIN32
+
+#include <rsutils/os/hresult.h>
+#include <KnownFolders.h>
+#include <shlobj.h>
+
+#elif defined __linux__ || defined __APPLE__
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+
+#endif
+
+
+namespace rsutils {
+namespace os {
+
+
+std::string get_special_folder( special_folder f )
+{
+    std::string res;
+
+#ifdef _WIN32
+
+
+    if( f == special_folder::temp_folder )
+    {
+        char buf[MAX_PATH+2];  // "The maximum possible return value is MAX_PATH+1"
+        auto len = GetTempPathA( sizeof( buf ), buf );  // "return the length of the string, not including the terminating null character"
+        if( len > 0 )  // zero on failure
+            res.assign( buf, len );
+        else
+            throw std::runtime_error( "failed to get temp folder" );
+    }
+    else
+    {
+        GUID folder;
+        switch( f )
+        {
+        case special_folder::user_desktop:
+            folder = FOLDERID_Desktop;
+            break;
+        case special_folder::user_documents:
+            folder = FOLDERID_Documents;
+            // The user's Documents folder location may get overridden, as we know OneDrive does in certain
+            // circumstances. In such cases, the new function, SHGetKnownFolderPath, does not always return the new
+            // path, while the deprecated function does.
+            CHAR path[MAX_PATH];
+            CHECK_HR( SHGetFolderPathA( NULL, CSIDL_PERSONAL, NULL, 0, path ) );
+            res = path;
+            res += "\\";
+            return res;
```

Comment:
I don't like the `return` here, it is not obvious.
If this is a special case that needs different handling then do it in an `if` like the temp folder

Reply:
Ahh, I see the code was moved from another location. Still, consider changing...

Reply:
Changed.